### ISSUE TITLE
 Can differential privacy's protective effect be verified？

### ISSUE THREAD

**Issue by MrLinNing** [privacytrustlab/ml_privacy_meter#108]
Your work is excellent, providing a great verification tool for security and privacy researchers. I would like to inquire whether your method can be combined with existing differential privacy defense frameworks, such as the [Opacus](https://github.com/pytorch/opacus) differential privacy framework. Is it possible to create a tutorial to demonstrate how to verify the effectiveness of differential privacy in defending against your MIA attack method? Thank you!

**Comment by MrLinNing** [comment]
Additionally, there is a puzzling issue in [this tutorial](https://github.com/privacytrustlab/ml_privacy_meter/blob/master/tutorials/white_box_attack.ipynb). For the CIFAR-10 dataset, although the training accuracy is relatively high, at over 80%, the testing accuracy is quite poor, at less than 50%. This is an overfitting phenomenon, and the model has no practical value. Suppose we want to increase the test accuracy by changing the training structure or hyperparameters (learning rate, batch size), the resulting MIA ROC is almost the same as random guessing. In this case, it seems that the MIA attack becomes meaningless. How should we understand this situation?

**Comment by taojiashu** [comment]
Hi @MrLinNing, we have published a new version of the tool that allows verifying the DP parameters of DP training/models. You can refer to this [file](https://github.com/privacytrustlab/ml_privacy_meter/blob/master/documentation/dpaudit.md) for instructions. The entire tool has been overhauled and now uses the state-of-the-art MIA, RMIA, as its testing backbone. You can try some of the examples we provided in this version.

**Comment by dpkpathak** [comment]
Hi @taojiashu, I’m trying to understand how to verify the offline settings in the current implementation. I reviewed the run_mia.py [file](https://github.com/privacytrustlab/ml_privacy_meter/blob/master/run_mia.py), but I’m unclear on how the data splits for reference model training are separated from the single target model. Could you provide some clarification? Thanks in advance!

**Comment by taojiashu** [comment]
The current implementation is based on RMIA. Given a dataset, half of the data will be randomly sampled to train the target model. For each pair of reference models, a fresh round of sampling will be conducted to select half of the dataset for training each reference model. For example, `reference_model_0_0` trains on data `[1,3,5]`, which are randomly selected, while `reference_model_0_1` trains on the other half, which is `[2,4,6]`. In this way, each data point would have an equal number of reference models trained/not trained on it. In the offline attack, we only use reference models that are not trained on the data point.
The particular functional call is here https://github.com/privacytrustlab/ml_privacy_meter/blob/71fb4d07e3392e1adc9746cb005f693b4415924e/run_mia.py#L87
The implementation is here
https://github.com/privacytrustlab/ml_privacy_meter/blob/71fb4d07e3392e1adc9746cb005f693b4415924e/models/utils.py#L298
Basically we create a Boolean split matrix where each row represents the model's usage of a particular point in training. Hope it explains.

**Comment by dpkpathak** [comment]
Thank you for your detailed response. However, I still find some discrepancies compared to your explanation.

The target model is defined as the model trained on the first dataset returned from:
https://github.com/privacytrustlab/ml_privacy_meter/blob/71fb4d07e3392e1adc9746cb005f693b4415924e/run_mia.py#L87

Later, during the calculation of the out signal, a reference model trained on the out world dataset is required, as seen here:
https://github.com/privacytrustlab/ml_privacy_meter/blob/71fb4d07e3392e1adc9746cb005f693b4415924e/attacks.py#L34

I have a concern regarding whether the reference model’s signal truly comes only from the out world dataset. While debugging with a single reference model (using the default CIFAR-10 configuration), I observed that the system trains four models:

The first model is treated as the target model.

However, models 2 and 3 are still used as reference models.

According to the split_dataset_for_training method, the third model and the target model share some overlap in training data, which seems incorrect in a practical attack setting. Could you confirm if this behavior is intended?

Additionally, I noticed that in
https://github.com/privacytrustlab/ml_privacy_meter/blob/71fb4d07e3392e1adc9746cb005f693b4415924e/attacks.py#L132,
the probability Pr(z) is being approximated differently compared to the algorithm in the RMIA paper. Could you clarify the reasoning behind this approach?

**Comment by dpkpathak** [comment]
Building on my previous comment, I have a further question: Why is separate population dataset is utilized for auditing, rather than using samples used for training reference model?

**Comment by taojiashu** [comment]
To audit privacy, we should first set up a privacy game where the scope of privacy is clearly defined. The current examples in this tool is based on the most prevalent privacy game, _Membership Inference Game for Average Model and Record_ (Definition 3.1 in [this paper](https://arxiv.org/pdf/2111.09679)).Under this framework, we audit the privacy risk of a training algorithm by analyzing an “average” model created from a training set drawn from the underlying distribution. In the implementation of the tool, which is based on the implementation of LiRA and RMIA, two most popular MIA algorithms out there, we randomly select the training set for each model we train. It is true that there will be an overlap between training sets, but the OUT notion is measured on the point level. For an offline attack under this privacy game, we only use models that are not trained on the target point, which is half of the total models we train. Note that the you can change the number of target models in the config file, so that the privacy analysis is more accurate under this privacy game. If you look at how to compute the total number of models that need to be trained, it considers the number of target models and reference models. The reference models when attacking model_x and be used as another target model in another auditing trial. These are all done to follow the privacy game.

If you want to train OUT models with a disjoint set of data, the question is how do you realistically get a disjoint set of data that are also from the same data distribution. This privacy game will be similar to Def 3.2 in [this paper](https://arxiv.org/pdf/2111.09679), where there is only one target model and the members are thus already fixed. Privacy Meter does not audit privacy defined in this privacy game right now out of the box, as an additional dataset needs to be passed in.

Lastly, the probability Pr(z) is computed in the exactly the same way as the RMIA paper. We just rearranged the terms for efficiency and brevity of coding.
